spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/MessageConverterConfigurer.java

```diff
@@ -294,16 +295,78 @@ private ContentTypeConvertingInterceptor(String contentType, boolean input) {
+		private final DefaultContentTypeResolver contentTypeResolver = new DefaultContentTypeResolver();
+		private final CompositeMessageConverterFactory converterFactory = new CompositeMessageConverterFactory();
 
 		@Override
 		public Message<?> preSend(Message<?> message, MessageChannel channel) {
-			MessageValues newMessageValues = MessageSerializationUtils.deserializePayload(new MessageValues(message), this.contentTypeResolver);
-			message = MessageBuilder.withPayload(newMessageValues.getPayload()).copyHeaders(newMessageValues.getHeaders()).build();
+			Class<?> targetClass = null;
+			MessageConverter converter = null;
+			MimeType contentType = message.getHeaders().containsKey(BinderHeaders.BINDER_ORIGINAL_CONTENT_TYPE)
+						? MimeType.valueOf((String)message.getHeaders().get(BinderHeaders.BINDER_ORIGINAL_CONTENT_TYPE))
+								: contentTypeResolver.resolve(message.getHeaders());
+
+			if (contentType != null){
+				if (contentType.equals(MessageConverterUtils.X_JAVA_SERIALIZED_OBJECT) ||
+						(contentType.getType().equals(MessageConverterUtils.X_JAVA_OBJECT.getType())
+								&& contentType.getSubtype().equals(MessageConverterUtils.X_JAVA_OBJECT.getSubtype()))){
+					// for Java and Kryo de-serialization we need to reset the content type
+					message = MessageBuilder.fromMessage(message).setHeader(MessageHeaders.CONTENT_TYPE, contentType).build();
+					converter = contentType.equals(MessageConverterUtils.X_JAVA_SERIALIZED_OBJECT)
+							? converterFactory.getMessageConverterForType(contentType)
+									: converterFactory.getMessageConverterForAllRegistered();
+					String targetClassName = contentType.getParameter("type");
+					if (StringUtils.hasText(targetClassName)) {
+						try {
+							targetClass = Class.forName(targetClassName, false, Thread.currentThread().getContextClassLoader());
+						}
+						catch (Exception e) {
+							throw new IllegalStateException("Failed to determine class name for contentType: "
+									+ message.getHeaders().get(BinderHeaders.BINDER_ORIGINAL_CONTENT_TYPE), e);
+						}
+					}
+				}
+			}
+
+			if (converter != null){
+				if (contentType.equals(MessageConverterUtils.X_JAVA_OBJECT) && targetClass == null){
+					throw new IllegalStateException("Can not deserialize into message since 'contentType` has not "
+							+ "being encoded with the actual target type."
+							+ "Consider 'application/x-java-object; type=foo.bar.MyClass'");
+				}
+				Object payload = converter.fromMessage(message, targetClass);
+				contentType = targetClass == null ? MessageConverterUtils.X_JAVA_SERIALIZED_OBJECT : MessageConverterUtils.X_JAVA_OBJECT;
+				message = MessageBuilder.withPayload(payload)
+						.copyHeaders(message.getHeaders())
+						.setHeader(MessageHeaders.CONTENT_TYPE, contentType)
+						.removeHeader(BinderHeaders.BINDER_ORIGINAL_CONTENT_TYPE)
+						.build();
+			}
+			else {
+				contentType = contentType == null ? contentTypeResolver.resolve(message.getHeaders()) : contentType;
```

Comment:
Probably the reassignment of `contentType` is not needed here. Only need the assignment if it is null.